File: Educational DP/F.py
s = input()
t = input()

dp = [[0 for _ in range(len(t) + 1)] for _ in range(len(s) + 1)]

for i in range(1, len(s) + 1):
    for j in range(1, len(t) + 1):
        if s[i - 1] == t[j - 1]:
            dp[i][j] = dp[i - 1][j - 1] + 1
        else:
            dp[i][j] = max(dp[i][j - 1], dp[i - 1][j])

    res = []    
    def buildSequence():
        i = len(dp) - 1
        j = len(dp[0]) - 1
        while i != 0 and j != 0:
            if dp[i][j] == dp[i - 1][j]:
                i -= 1
            elif dp[i][j] == dp[i][j - 1]:
                j -= 1
            else:   
                res.append(s[i - 1])
                i -= 1
                j -= 1
    
    buildSequence()
    res.reverse()
    output = "".join(res)
    print(output)

File: EDPC/F.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
  string s, t;
  cin >> s >> t;
  int a, b;
  a = s.size() + 1;
  b = t.size() + 1;
  vector lcs(a, vector<int>(b, 0));

  for (int i = 1; i < a; i++) {
    for (int j = 1; j < b; j++) {
      if (s[i - 1] == t[j - 1]) lcs[i][j] = lcs[i-1][j-1] + 1;
      else lcs[i][j] = max(lcs[i-1][j], lcs[i][j-1]);
    }
  }

  int i = a-1;
  int j = b-1;
  string res = "";
  while (i > 0 && j > 0) {
    if (lcs[i][j] == lcs[i-1][j]) i--;
    else if (lcs[i][j] == lcs[i][j-1]) j--;
    else {
      res += s[i-1];
      i--;
      j--;
    }
  }
  reverse(res.begin(), res.end());
  cout << res << endl;
}
